include/Oscillator.h

```diff
@@ -153,6 +165,21 @@ class EXPORT Oscillator
 		return m_userWave->userWaveSample( _sample );
 	}
 
+	inline sample_t wtSample(const WaveShapes shape, const float _sample)
+	{
+		const float frame = _sample * WAVETABLE_LENGTH;
+		f_cnt_t f1 = static_cast<f_cnt_t>(frame) % WAVETABLE_LENGTH;
+		if (f1 < 0)
+		{
+			f1 += WAVETABLE_LENGTH;
+		}
+		f_cnt_t f2 = f1 < WAVETABLE_LENGTH - 1 ?
+					f1 + 1 :
+					0;
+		return linearInterpolate(s_waveTables[shape][waveTableBandFromFreq(m_freq)][f1],
+				s_waveTables[shape][waveTableBandFromFreq(m_freq)][f2], fraction(frame));
```

Comment:
Right now, the compiler can't always see into `waveTableBandFromFreq(m_freq)` since that's defined in the source file. So I'll bet it's going to actually call that function twice, since it won't be able to verify it has no side-effects. This _might_ change if you move that function into the header, but I'd feel more comfortable if you assigned that result to a variable, e.g. something like

```
int band = waveTableBandFromFreq(m_freq);
return linearInterpolate(s_waveTables[shape][band][f1], s_waveTables[shape][band][f2], fraction(frame));
```

Reply:
@curlymorphic I don't know any specific resources to point you to regarding compiler optimizations. If you can read assembly code, you can figure out what kinds of things compilers are capable of by typing code into http://godbolt.org and then trying to figure out what the assembly code is doing. You can even select old compiler versions and see how the code they generate and the optimization they can do has improved over time. But that can be a lot of work.

Reply:
@curlymorphic The optimization @Wallacoloo is describing is called [common subexpression elimination](https://en.wikipedia.org/wiki/Common_subexpression_elimination).